### PR TITLE
7131 caching of node js compilation on model bundling

### LINKED ISSUE #7131: Prove of concept: Caching of nodeJS compilation on model bundling
Dear Bokeh developers!

A while ago we shortly discussed on Gitter that in Bokeh server the compile time to generate the custom model JS inline is slow and should be cached.
Here we present a simple prove of concept for a caching of the nodeJS compilation that holds even when more than one disjunct models were delivered by a single bokeh server instance.

In bokeh/bokeh/util/compiler.py
```
def calc_cache_key():
    """There is no metadata than the Model classes,
        so this is the only base to generate a cache  key.
       We build the model keys from the list of model.full_name(s).
       This is ineffective but possibly a better solution can be found found.
       It is still at least an order of magnitude faster faster than the nodeJS :-)
    """

    models = Model.model_class_reverse_map.values()
    custom_model_names = ""

    for cls in models:
        impl = getattr(cls, "__implementation__", None)

        if impl is not None:
            model = CustomModel(cls)
            custom_model_names += model.full_name

    key = hashlib.sha256(custom_model_names.encode('utf-8')).hexdigest()
    return key

def bundle_all_models():
    if not hasattr(Model, '__cache'):
        Model.__cache = {}
    key = calc_cache_key()
    if key in Model.__cache:
        return Model.__cache[key]
    else:
        Model.__cache[key] = bundle_models(Model.model_class_reverse_map.values()) or ""
        return Model.__cache[key]
    if not Model.__cache[key]: #this is probably overkill
        Model.__cache[key] = bundle_models(Model.model_class_reverse_map.values()) or ""
    return Model.__cache[key]

#    return bundle_models(Model.model_class_reverse_map.values()) or ""
```
Cheers,
Volker

### PR DESCRIPTION
**Pull request by volkerjaenisch**
- [ ] issues: fixes #7131

### CODE REVIEW

File: bokeh/util/compiler.py [hunk 1]
```diff
@@ -426,5 +426,36 @@ def resolve_deps(deps, root):
...
+        Model.__cache = {}
+    key = calc_cache_key()
+    if key in Model.__cache:
+        return Model.__cache[key]
+    else:
+        Model.__cache[key] = bundle_models(Model.model_class_reverse_map.values()) or ""
+        return Model.__cache[key]
+    if not Model.__cache[key]: #this is probably overkill
+        Model.__cache[key] = bundle_models(Model.model_class_reverse_map.values()) or ""
+    return Model.__cache[key]
```
[the remark this answers is not in the document]
Comment (volkerjaenisch) [reply bokeh/util/compiler.py]: I will change this.

[the remark this answers is not in the document]
Comment (volkerjaenisch) [reply bokeh/util/compiler.py]: How can I update the pull request about the new commit in the branch?

File: bokeh/util/compiler.py [hunk 2]
```diff
@@ -447,15 +447,11 @@ def calc_cache_key():
...
-        return Model.__cache[key]
-    else:
-        Model.__cache[key] = bundle_models(Model.model_class_reverse_map.values()) or ""
-        return Model.__cache[key]
-    if not Model.__cache[key]: #this is probably overkill
-        Model.__cache[key] = bundle_models(Model.model_class_reverse_map.values()) or ""
-    return Model.__cache[key]
+    bundle = _bundle_cache.get(key, None)
+    if bundle is None:
+        _bundle_cache[key] = bundle = bundle_models(...) or ""
```
Comment (mattpap) [inline bokeh/util/compiler.py]: @volkerjaenisch, those `...` need to be filled in. I didn't write this down for brevity.

File: bokeh/util/compiler.py [hunk 3]
```diff
@@ -447,15 +447,11 @@ def calc_cache_key():
...
+    return bundle
+
```
Comment (mattpap) [inline bokeh/util/compiler.py]: Empty trailing line will cause code quality tests to fail.

Comment (volkerjaenisch) [reply bokeh/util/compiler.py]: Sorry, was distracted. Too many projects at a time.

**Comment by bryevdv** [comment]
@mattpap any idea why CI is not running for this PR?

**Comment by mattpap** [comment]
@bryevdv, no idea, but I've been seeing issues with github's connectivity to external services in other PRs recently. @volkerjaenisch, can you squash your changes and repush? Maybe this will help.

**Comment by volkerjaenisch** [comment]
done.

**Comment by bryevdv** [comment]
Still not triggering CI. I will check out the branch by hand today.

**Comment by bryevdv** [comment]
There's actually a formatting issue in the docstring that causes docs fail. Also the dumb explicit re-ordering that put extensions examples last because of the slowness can be removed. I will have a patch for you soon.

**Comment by bryevdv** [comment]
@volkerjaenisch can you apply this diff? Then I will merge:
```
diff --git a/bokeh/sphinxext/bokeh_plot.py b/bokeh/sphinxext/bokeh_plot.py
index c4e2d7900..b297f7d9a 100644
--- a/bokeh/sphinxext/bokeh_plot.py
+++ b/bokeh/sphinxext/bokeh_plot.py
@@ -265,7 +265,7 @@ class BokehPlotDirective(Directive):
         return result

 def env_before_read_docs(app, env, docnames):
-    docnames.sort(key=lambda x: 2 if "extension" in x else 0 if "examples" in x else 1)
+    docnames.sort(key=lambda x: 0 if "examples" in x else 1)
     for name in [x for x in docnames if env.doc2path(x).endswith(".py")]:
         if not name.startswith(tuple(env.app.config.bokeh_plot_pyfile_include_dirs)):
             env.found_docs.remove(name)
diff --git a/bokeh/util/compiler.py b/bokeh/util/compiler.py
index f08818507..126caeff3 100644
--- a/bokeh/util/compiler.py
+++ b/bokeh/util/compiler.py
@@ -427,12 +427,15 @@ def bundle_models(models):
     return _plugin_umd % dict(content=content)

 def calc_cache_key():
-    """There is no metadata than the Model classes,
-        so this is the only base to generate a cache  key.
-       We build the model keys from the list of model.full_name(s).
-       This is ineffective but possibly a better solution can be found found.
-       It is still at least an order of magnitude faster faster than the nodeJS :-)
-    """
+    ''' Generate a key to cache a custom extension impementation with.
+
+    There is no metadata other than the Model classes, so this is the only
+    base to generate a cache key.
+
+    We build the model keys from the list of ``model.full_name``. This is
+    not ideal but possibly a better solution can be found found later.
+
+    '''

     models = Model.model_class_reverse_map.values()
     custom_model_names = ""
```

File: bokeh/util/compiler.py [hunk 4]
```diff
@@ -426,5 +426,31 @@ def resolve_deps(deps, root):
...
+    return key
+
+_bundle_cache = {}
 def bundle_all_models():
-    return bundle_models(Model.model_class_reverse_map.values()) or ""
+    key = calc_cache_key()
+    bundle = _bundle_cache.get(key, None)
+    if bundle is None:
+        _bundle_cache[key] = bundle = bundle_models(Model.model_class_reverse_map.values()) or ""
+    return bundle
```
Comment (mattpap) [inline bokeh/util/compiler.py]: No trailing newline will cause code quality tests to fail.

**Comment by bryevdv** [comment]
@mattpap I know, but I have no idea how to make them run. I tested the PR and docs build locally, anything else build related will have to get fixed in a follow up PR

**Comment by mattpap** [comment]
OK. Just checking. Given recent flakiness of travis ci integration, I would prefer to have this comment at merge time to avoid confusion.

**Comment by bryevdv** [comment]
Fair enough I will add one if a similar situation arises again (FYI this merge did end up passing)

**Pull request merged**